frontend/lib/src/components/widgets/NumberInput/NumberInput.tsx

```diff
@@ -52,510 +52,441 @@ import {
   }
 
-  private getMin = (): number => {
-    return this.props.element.hasMin ? this.props.element.min : -Infinity
-  }
-
-  private getMax = (): number => {
-    return this.props.element.hasMax ? this.props.element.max : +Infinity
-  }
-
-  private getStep = (): number => {
-    const { step } = this.props.element
-
-    if (step) {
-      return step
-    }
-    if (this.isIntData()) {
-      return 1
-    }
-    return 0.01
+  const updateFromProtobuf = (): void => {
+    const { value } = element
+    element.setValue = false
+    setValue(value ?? null)
+    setFormattedValue(formatValue({ value: value ?? null, ...element, step }))
+    commitValue({ value: value ?? null, source: { fromUi: false } })
   }
 
-  /** Commit state.value to the WidgetStateManager. */
-  private commitWidgetValue = (source: Source): void => {
-    const { value } = this.state
-    const { element, widgetMgr, fragmentId } = this.props
-    const data = this.props.element
-
-    const min = this.getMin()
-    const max = this.getMax()
-
-    if (notNullOrUndefined(value) && (min > value || value > max)) {
-      const node = this.inputRef.current
-      if (node) {
-        node.reportValidity()
-      }
+  // on component mount, we want to update the value from protobuf if setValue is true, otherwise commit current value
+  React.useEffect(() => {
+    const formClearHelperCopy = formClearHelper.current
+    if (element.setValue) {
+      updateFromProtobuf()
     } else {
-      const valueToBeSaved = value ?? data.default ?? null
-
-      if (this.isIntData()) {
-        widgetMgr.setIntValue(element, valueToBeSaved, source, fragmentId)
-      } else {
-        widgetMgr.setDoubleValue(element, valueToBeSaved, source, fragmentId)
-      }
+      commitValue({ value, source: { fromUi: false } })
+    }
 
-      this.setState({
-        dirty: false,
-        value: valueToBeSaved,
-        formattedValue: this.formatValue(valueToBeSaved),
-      })
+    return () => {
+      formClearHelperCopy.disconnect()
     }
-  }
 
-  /**
-   * If we're part of a clear_on_submit form, this will be called when our
-   * form is submitted. Restore our default value and update the WidgetManager.
-   */
-  private onFormCleared = (): void => {
-    this.setState(
-      (_, prevProps) => {
-        return { value: prevProps.element.default ?? null }
-      },
-      () => this.commitWidgetValue({ fromUi: true })
-    )
-  }
+    // I don't want to run this effect on every render, only on mount.
+    // Additionally, it's okay if commitValue changes, because we only call
+    // it once in the beginning anyways.
+    /* eslint-disable react-hooks/exhaustive-deps */
+  }, [])
 
-  private onBlur = (): void => {
-    if (this.state.dirty) {
-      this.commitWidgetValue({ fromUi: true })
+  // update from protobuf whenever component updates if element.setValue is truthy
+  React.useEffect(() => {
+    if (element.setValue) {
+      updateFromProtobuf()
     }
+  })
```

Comment:
I believe leaving out the dependency array runs the `useEffect` hook every time which means it could be left out completely (see [docs](https://react.dev/reference/react/useEffect#my-effect-runs-after-every-re-render)).

Reply:
good point!